File: flutter/packages/hatake_yaml/test/conformance_cross_field_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:hatake_core/hatake_core.dart';
import 'package:hatake_yaml/hatake_yaml.dart';
import 'package:test/test.dart';

/// Runs the shared cross-field-validation fixture
/// (spec/conformance/cross_field_validation.json) — the same contract the
/// TypeScript and Java editions implement.
///
/// ここで固定するのは「相手の項目と比べる」「数として読めれば数・読めなければ文字」
/// 「`aggregate` で明細を畳んだ数と比べる」「どちらかが空なら通す」「メッセージは
/// 相手のラベルで出す」の5つ。フロントとバックで検証がズレないことが値打ちなので、
/// 同じフィクスチャを3エディションで回す。
void main() {
  final fixture = jsonDecode(
    File('../../../spec/conformance/cross_field_validation.json')
        .readAsStringSync(),
  ) as Map<String, dynamic>;

  // strict で読む＝フィクスチャが「本当に書ける定義」であることも縛る。
  final page = parsePageJson(jsonEncode(fixture['page']), strict: true)
      as FormPageDefinition;
  final validator = FormValidator();

  group('conformance: cross-field validation', () {
    for (final raw in fixture['cases'] as List) {
      final c = raw as Map<String, dynamic>;
      test(c['name'], () {
        final record = (c['record'] as Map).cast<String, Object?>();
        final actual = validator
            .validate(page.form, record, mode: c['mode'] as String?)
            .errors;

        String key(String field, String message) => '$field=$message';
        final expected = [
          for (final e in c['expected'] as List)
            key((e as Map)['field'] as String, e['message'] as String),
        ];
        expect(
          actual.map((e) => key(e.field, e.message)).toSet(),
          expected.toSet(),
          reason: 'errors: ${actual.map((e) => key(e.field, e.message))}',
        );
      });
    }
  });
}

File: typescript/test/conformanceCrossField.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import {
  FormValidator,
  parsePageJson,
  type FormPageDefinition,
  type ValidationError,
} from "../src/index.js";

// 項目間の検証（compare）と、**どれを先に言うか**の共有フィクスチャ。Dart 版・Java 版も
// 同じものを食べる（検証がフロントとバックでズレないことが、この DSL の値打ちなので）。
const key = (e: ValidationError | { field: string; message: string }): string =>
  `${e.field}=${e.message}`;

function runFixture(name: string, file: string): void {
  const fixture = JSON.parse(readFileSync(`../spec/conformance/${file}`, "utf8"));

  describe(`conformance: ${name}`, () => {
    // strict で読む＝フィクスチャが「本当に書ける定義」であることも、ここで縛る。
    const page = parsePageJson(JSON.stringify(fixture.page), {
      strict: true,
    }) as FormPageDefinition;
    const validator = new FormValidator();

    for (const c of fixture.cases as any[]) {
      it(c.name, () => {
        const actual = validator.validate(page.form, c.record, c.mode).errors;
        expect(new Set(actual.map(key))).toEqual(
          new Set((c.expected as any[]).map(key)),
        );
      });
    }
  });
}

runFixture("cross-field validation", "cross_field_validation.json");
// 1項目で複数落ちたとき、自分の形が先・他の項目に依るものが後。
runFixture("which error is reported first", "validation_order.json");
